### ISSUE TITLE
Qwen2-VL output diff b/w huggingface and Tensorrt-LLM

### ISSUE THREAD

**Issue by Ericoool9614** [NVIDIA/TensorRT-LLM#3235]
same model/weights/environment/sampling params and so on. But the output of Tensorrt-LLM engine is different from huggingface one.
anyone have the similar problem?

**Comment by hchings** [comment]
Could you share a minimal reproducible commands and outputs? 

Even with the same weights, minor differences can occur due to differences in default configs (e.g., sampling config).

That said, the outputs should still make sense. If they're drastically different or incorrect, we’d definitely want to investigate further.

**Comment by Ericoool9614** [comment]
> Could you share a minimal reproducible commands and outputs?
> 
> Even with the same weights, minor differences can occur due to differences in default configs (e.g., sampling config).
> 
> That said, the outputs should still make sense. If they're drastically different or incorrect, we’d definitely want to investigate further.

Sure.
Enviroment:
Tensorrt-LLM 0.16.0
L20
Python 3.10.12
Model is quantized via autoawq, base model is Qwen2-vl-7b-instruct

Step 1: Convert checkpoint
python ../qwen/convert_checkpoint.py --model_dir /home/hzh/Qwen2-VL-7B-Instruct-AWQ --output_dir /home/hzh/trt_llm_ckpt/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2 --dtype float16

Step 2: build engine
trtllm-build --checkpoint_dir /home/hzh/trt_llm_ckpt/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2 --output_dir /home/hzh/trt_llm_engine/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2 --max_batch_size=1 --max_input_len=8192 --max_seq_len=8192 --max_multimodal_len=8192 --gemm_plugin=float16 --gpt_attention_plugin=float16 --use_paged_context_fmha enable

Step 3: build visual engine
python build_visual_engine.py --model_type qwen2_vl --model_path /home/hzh/Qwen2-VL-7B-Instruct-AWQ --output_dir /home/hzh/trt_llm_engine/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2-vision

Step 4: 
trtllm inference, using examples/multimodal/run.py
python run.py --hf_model_dir /home/hzh/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2 --visual_engine_dir /home/hzh/trt_llm_engine/Qwen2-VL-7B-Instruct-mllm-chinese-vision  --llm_engine_dir /home/hzh/trt_llm_engine/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2 --max_new_tokens 300 --top_k 1 --top_p 0.01 --temperature 0 --repetition_penalty 1.1 --input_text "You are a professional summary expert. You generate a summary based on the image content. Association and diffusion are prohibited, and guessing the relationship between characters is prohibited. Limit 100 words" --image_path "/home/hzh/qwenvl/img_data/row2/img_2.jpeg"

output from Tensorrt-LLM:
"The image is a black-and-white portrait of a young man in military attire, including a cap with a star emblem. The expression is neutral, and the background is plain, focusing attention solely on the subject. This style of dress and the cap suggest a historical or military context, possibly from mid-20th century East Asian or Soviet military uniforms. There are no other people or objects in the image, emphasizing the individual's presence."

output from huggingface(using same sampling params):
"The image features a historical black-and-white photograph of a young man in military attire, specifically wearing a cap with a star emblem and a uniform that appears to be from mid-20th century Chinese or Soviet military style. The expression on his face is neutral, and he gazes directly at the camera, suggesting a formal portrait rather than an action shot. There are no visible signs of movement or interaction with other subjects, indicating this might have been taken during a moment of rest or posed for the photograph. The background is plain, ensuring focus remains solely on the individual. This type of imagery could be used in educational contexts about military history or personal stories related to that era."

**Comment by hchings** [comment]
Hi @Ericoool9614, thanks. Would you mind to share your huggingface script as well for easier repro? I'm trying it with the latest TensorRT-LLM without autoawq first.

**Comment by Ericoool9614** [comment]
> Hi [@Ericoool9614](https://github.com/Ericoool9614), thanks. Would you mind to share your huggingface script as well for easier repro? I'm trying it with the latest TensorRT-LLM without autoawq first.

`
# Load model directly
from transformers import AutoProcessor, Qwen2VLForConditionalGeneration,GenerationConfig
from transformers import AutoTokenizer
from PIL import Image
import requests
import pandas as pd
import torch 
import os
import json
import time
import base64
from openpyxl import Workbook, load_workbook
from openpyxl.drawing.image import Image as OpenpyxlImage
from qwen_vl_utils import process_vision_info


def get_mm_prompt(processor, image_path):
    conversation = [
        {
            "role":"user",
            "content":[
                {
                    "type": "image",
                    "image": Image.open(image_path)
                },
                {
                    "type": "text",
                    "text": "You are a professional summary expert. You generate a summary based on the image content. Association and diffusion are prohibited, and guessing the relationship between characters is prohibited. Limit 100 words"
                }
            ]
        }
    ]
    text = processor.apply_chat_template(conversation, tokenize=False, add_generation_prompt=True)
    image_inputs, video_inputs = process_vision_info(conversation)
    inputs = processor(
        text=[text],
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors="pt",
    )
    inputs = inputs.to("cuda")

    return inputs

if __name__ == '__main__':
    model_path = "/home/hzh/Qwen2-VL-7B-Instruct-mllm-chinese-awq-30data-pt2"
    processor = AutoProcessor.from_pretrained(model_path)
    model = Qwen2VLForConditionalGeneration.from_pretrained(
        model_path,
        torch_dtype=torch.float16,
        attn_implementation="flash_attention_2",
        device_map="auto",
        ).eval().cuda()

    inputs = get_mm_prompt(processor, "/home/hzh/qwenvl/img_data/row2/img_2.jpeg")
    generated_ids = model.generate(**inputs,
        max_new_tokens=300,
        #top_p=0.6, 
        #top_k=50,  
        repetition_penalty=1.1, 
        # temperature=0.0,  
        do_sample=False 
        )
    generated_ids_trimmed = [
        out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
    ]
    output_text = processor.batch_decode(
        generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
    )
    # print(output_text)
    res = ""
    try:
        print(json.loads(output_text[0]))
        res = str(json.loads(output_text[0]))
    except:
        print(str(output_text[0]))
        res = str(output_text[0])
`

**Comment by Ericoool9614** [comment]
> Hi [@Ericoool9614](https://github.com/Ericoool9614), thanks. Would you mind to share your huggingface script as well for easier repro? I'm trying it with the latest TensorRT-LLM without autoawq first.

Hi @hchings ! May I ask if this issue got any update? Is it useful if I change to the newest pytorch flow in v0.19.0rc0?

**Comment by hchings** [comment]
Hi @Ericoool9614, I tried w/ TRT-LLM `0.20.0rc1` and `Qwen2-VL-7B-Instruct` (without `AWQ`) on several images, the outputs from TRT-LLM (below left) look semantically correct to me but they're not identical to HuggingFace (HF, below right)'s. 

<img width="1430" alt="Image" src="https://github.com/user-attachments/assets/a210dc2d-5352-4790-a3f9-35cf94be5b82" />

I don't think TRT-LLM should be expected to have identical output as HF. 
Some possible sources that could cause the diff that I can think of so far:
- TRT-LLM's `MultimodalEngineBuilder`'s image pre-processing step (e.g., [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/tools/multimodal_builder.py#L1357)) is not identical to your HF script.
- It seems that even set `do_sample=False` and unset `temperature`, etc, HF is still using `"temperature": 0.01` and `"do_sample": true` in its `logits_processor`([here](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2383)). 
    - I didn't spend too much time digging into how HF determine their generation config or have which implicit sampling logics, but from initial check, they don't look identical to TRT-LLM's. 
- TRT-LLM might have some fusion ops which may affect accuracy.
- TRT-LLM multimodal script is not using FlashAttention/FlashAttention2, but HF script does. This theoretically shouldn't impact accuracy, but I'm not sure whether there's any nuances there that might slightly affect the outputs.

Overall It doesn't look like a bug to me in TRT-LLM and you're likely to sill see slightly diff outputs w/ latest version. I can have the team to investigate further by filing a bug, but it's likely a lower priority issue for us so it might take a while, given the outputs still look correct.
